src/protocol/sort/shuffle.rs

```diff
@@ -0,0 +1,280 @@
+use crate::helpers::prss::PrssSpace;
+use rand::rngs::mock::StepRng;
+use rand::seq::SliceRandom;
+
+use embed_doc_image::embed_doc_image;
+use futures::future::try_join_all;
+
+use crate::{
+    error::BoxError,
+    field::Field,
+    helpers::{
+        mesh::{Gateway, Mesh},
+        prss::SpaceIndex,
+        Direction, Identity,
+    },
+    protocol::{context::ProtocolContext, RecordId, Step},
+    secret_sharing::Replicated,
+};
+
+use super::{
+    apply::apply_inv_inplace,
+    reshare::Reshare,
+    ShuffleStep::{self, Step1, Step2, Step3},
+};
+
+#[allow(dead_code)]
+pub struct Shuffle<F, S> {
+    input: Vec<Replicated<F>>,
+    step_fn: fn(ShuffleStep) -> S,
+}
+
+/// This is SHUFFLE(Algorithm 1) described in <https://eprint.iacr.org/2019/695.pdf>.
```

Comment:
Do we want to update the linked paper to the new one?

Reply:
No, all these algorithms are being implemented from the older semi honest paper. We will just pickup the malicious algorithms from the new paper when we reach that point.